File: config/src/main/java/com/typesafe/config/impl/MemoKey.java
package com.typesafe.config.impl;

/** The key used to memoize already-traversed nodes when resolving substitutions */
final class MemoKey {
    MemoKey(AbstractConfigValue root, AbstractConfigValue value, Path restrictToChildOrNull) {
        this.root = root;
        this.value = value;
        this.restrictToChildOrNull = restrictToChildOrNull;
    }

    final private AbstractConfigValue root;
    final private AbstractConfigValue value;
    final private Path restrictToChildOrNull;

    @Override
    public final int hashCode() {
        int h = System.identityHashCode(value);
        h = h + 41 * (41 + root.hashCode());
        if (restrictToChildOrNull != null) {
            return h + 41 * (41 + restrictToChildOrNull.hashCode());
        } else {
            return h;
        }
    }

    @Override
    public final boolean equals(Object other) {
        if (other instanceof MemoKey) {
            MemoKey o = (MemoKey) other;
            if (o.value != this.value)
                return false;
            else if (o.root != this.root)
                return false;
            else if (o.restrictToChildOrNull == this.restrictToChildOrNull)
                return true;
            else if (o.restrictToChildOrNull == null || this.restrictToChildOrNull == null)
                return false;
            else
                return o.restrictToChildOrNull.equals(this.restrictToChildOrNull);
        } else {
            return false;
        }
    }

    @Override
    public final String toString() {
        return "MemoKey(" + value + "@" + System.identityHashCode(value) + "," + restrictToChildOrNull + ")";
    }
}

File: config/src/main/scala/com/typesafe/config/impl/MemoKey.scala
package com.typesafe.config.impl

/** The key used to memoize already-traversed nodes when resolving substitutions */
final class MemoKey private[impl] (val value: AbstractConfigValue, val restrictToChildOrNull: Path) {

    override final def hashCode(): Int = {
        val h = System.identityHashCode(value)
        if (restrictToChildOrNull != null)
            h + 41 * (41 + restrictToChildOrNull.hashCode)
        else h
    }

    override final def equals(other: Any): Boolean =
        if (other.isInstanceOf[MemoKey]) {
            val o = other.asInstanceOf[MemoKey]
            if (o.value ne this.value) false
            else if (o.restrictToChildOrNull eq this.restrictToChildOrNull) true
            else if (o.restrictToChildOrNull == null || this.restrictToChildOrNull == null)
                false
            else o.restrictToChildOrNull == this.restrictToChildOrNull
        } else false

    override final def toString(): String =
        "MemoKey(" + value + "@" + System.identityHashCode(value) + "," + restrictToChildOrNull + ")"
}
